lib/migrate.js

```diff
@@ -0,0 +1,496 @@
+        results.indexesToDrop = oldIndexes;
+      } else {
+        results = self.compare(newIndexes, oldIndexes);
+      }
+      cb(null, results);
+    });
+
+    function generatePropertyLevelIndexes(indexes) {
+      var results = {};
+      for (var key in indexes) {
+        var field = {};
+        // By default the order will be `asc`,
+        // please create Model level index if you need `desc`
+        field[key.split('_index')[0]] = 'asc';
+        var fields = [field];
+        results[key] = fields;
+      };
+      return results;
+    }
+
+    function generateModelLevelIndexes(indexes) {
+      var results = {};
+      for (var key in indexes) {
+        var keys = indexes[key].keys;
+        if (!keys || typeof keys !== 'object') return cb(new Error(
+          'the keys in your model index are not well defined! please see' +
+          'https://loopback.io/doc/en/lb3/Model-definition-JSON-file.html#indexes'));
+
+        var fields = [];
+        _.forEach(keys, function(value, key) {
+          var obj = {};
+          var order;
+          if (keys[key] === 1) order = 'asc';
+          else order = 'desc';
+          obj[key] = order;
+          fields.push(obj);
+        });
+        results[key] = fields;
+      }
+      return results;
+    }
+  };
+
+  /**
+ * Perform the indexes comparison for `autoupdate`.
+ * @param {Object} newIndexes
+ * newIndexes in format:
+ * ```js
+ * {
+ *   indexName: [{afield: 'asc'}],
+ *   compositeIndexName: [{field1: 'asc'}, {field2: 'asc'}]
+ * }
+ * ```
+ * @param {Object} oldIndexes
+ * oldIndexes in format:
+ * ```js
+ * {
+ *   indexName: {
+ *     ddoc: '_design/LBModel__Foo__LBIndex__bar_index',
+ *     fields: [{afield: 'asc'}],
+ *   }
+ * }
+ * ```
+ * @callback {Function} cb The callback function
+ * @param {Object} result indexes to add and drop after comparison
+ * ```js
+ * result: {indexesToAdd: {$someIndexes}, indexesToDrop: {$someIndexes}}
+ * ```
+ */
+  CouchDB.prototype.compare = function(newIndexes, oldIndexes, cb) {
+    debug('CouchDB.prototype.compare');
+    var result = {};
+    var indexesToDrop = {};
+    var indexesToAdd = {};
+    var iAdd;
+    for (var niKey in newIndexes) {
+      if (!oldIndexes.hasOwnProperty(niKey)) {
+      // Add item to `indexesToAdd` if it's new
+        iAdd = {};
+        iAdd[niKey] = newIndexes[niKey];
+        indexesToAdd = _.merge(indexesToAdd, iAdd);
+      } else {
+        if (arrEqual(newIndexes[niKey], oldIndexes[niKey].fields)) {
+        // Don't change it if index already exists
+          delete oldIndexes[niKey];
+        } else {
+        // Update index if fields change
+          iAdd = {};
+          var iDrop = {};
+          iAdd[niKey] = newIndexes[niKey];
+          indexesToAdd = _.merge(indexesToAdd, iAdd);
+        }
+
+        function arrEqual(arr1, arr2) {
+          if (!Array.isArray(arr1) || !Array.isArray(arr2))
+            return cb(new Error('make sure you pass in two Array!'));
```

Comment:
Thinking about it further, just return false if they don't give you sane input.
`{}` isn't equal to `[]`, so just say so. :P